qcodes/instrument_drivers/QDevil/QDevil_QDAC.py

```diff
@@ -0,0 +1,938 @@
+        handle.baud_rate = 480600
+        handle.parity = visa.constants.Parity(0)
+        handle.data_bits = 8
+        self.set_terminator('\n')
+        # TODO: do we want a method for write termination too?
+        handle.write_termination = '\n'
+        # TODO: do we need a query delay for robust operation?
+        self._write_response = ''
+        firmware_version = self._get_firmware_version()
+        if firmware_version < 1.07:
+            LOG.warning("Firmware version: {}".format(firmware_version))
+            raise RuntimeError('''
+                No QDevil QDAC detected or the firmware version is obsolete.
+                This driver only supports version 1.07 or newer. Please
+                contact info@qdevil.com for a firmware update.
+                ''')
+
+        self.num_chans = self._get_number_of_channels()
+        num_boards = int(self.num_chans/8)
+        self._output_n_lines = self.num_chans + 2
+
+        # Assigned slopes. Entries will eventually be {chan: slope}
+        self._slopes = {}
+        # Function generators and triggers (used in ramping)
+        self._fgs = set(range(1, 9))
+        self._assigned_fgs = {}  # {chan: fg}
+        self._trigs = set(range(1, 10))
+        self._assigned_triggers = {}  # {fg: trigger}
+        # Sync channels
+        self._syncoutputs = {}  # {chan: syncoutput}
+
+        self._chan_range = range(1, 1 + self.num_chans)
+        self.channel_validator = vals.Ints(1, self.num_chans)
+
+        channels = ChannelList(self, "Channels", QDacChannel,
+                               snapshotable=False,
+                               multichan_paramclass=QDacMultiChannelParameter)
+
+        for i in self._chan_range:
+            channel = QDacChannel(self, 'chan{:02}'.format(i), i)
+            channels.append(channel)
+            self.add_submodule('ch{:02}'.format(i), channel)
+        channels.lock()
+        self.add_submodule('channels', channels)
+
+        for board in range(num_boards):
+            for sensor in range(3):
+                label = 'Board {}, Temperature {}'.format(board, sensor)
+                self.add_parameter(name='temp{}_{}'.format(board, sensor),
+                                   label=label,
+                                   unit='C',
+                                   get_cmd='tem {} {}'.format(board, sensor),
+                                   get_parser=self._num_verbose)
+
+        self.add_parameter(name='cal',
+                           set_cmd='cal {}',
+                           vals=self.channel_validator)
+
+        self.add_parameter(name='verbose',
+                           label=label,
+                           val_mapping={True: 1, False: 0})
+
+        # Initialise the instrument, all channels DC (unbind func. generators)
+        for chan in self._chan_range:
+            # Note: this call may change the voltage on the channel
+            self.write('wav {} 0 1 0'.format(chan))
```

Comment:
I get the point. Of course "reset" method will be good to have in any case. But it is a bit of a complicated to read in the actual state including sync assignments, and generator assignment, especially if some generators are running either because they are ramping or because the user has started them directly by a "write" command. I don't think the current firmware can tell the remaining time for a ramp, but it can be calculated from the voltages. So it is not a noop to do thoroughly. I will think more about it.

Alternatively just freezing the outputs will be better than the present solution.

Reply:
Implemented a _load_state method called at initialisation and a reset method